crates/sui-indexer-alt-e2e-tests/src/lib.rs

```diff
@@ -1,6 +1,387 @@
+            with_genesis,
+            registry,
+            cancel.child_token(),
+        )
+        .await
+        .context("Failed to setup indexer")?;
+
+        let pipelines = indexer.pipelines().collect();
+        let indexer = indexer.run().await.context("Failed to start indexer")?;
+
+        let jsonrpc = start_rpc(
+            db_args,
+            rpc_args,
+            system_package_task_args,
+            rpc_config,
+            registry,
+            cancel.child_token(),
+        )
+        .await
+        .context("Failed to start JSON-RPC server")?;
+
+        Ok(Self {
+            rpc_listen_address,
+            db,
+            pipelines,
+            indexer,
+            jsonrpc,
+            database,
+            cancel,
+        })
+    }
+
+    /// The URL to talk to the database on.
+    pub fn db_url(&self) -> Url {
+        self.database.database().url().clone()
+    }
+
+    /// The URL to send JSON-RPC requests to.
+    pub fn rpc_url(&self) -> Url {
+        Url::parse(&format!("http://{}/", self.rpc_listen_address))
+            .expect("Failed to parse RPC URL")
+    }
+
+    /// Returns the latest checkpoint that we have all data for in the database, according to the
+    /// watermarks table. Returns `None` if any of the expected pipelines are missing data.
+    pub async fn latest_checkpoint(&self) -> anyhow::Result<Option<u64>> {
+        use watermarks::dsl as w;
+
+        let mut conn = self
+            .db
+            .connect()
+            .await
+            .context("Failed to connect to database")?;
+
+        let latest: HashMap<String, i64> = w::watermarks
+            .select((w::pipeline, w::checkpoint_hi_inclusive))
+            .filter(w::pipeline.eq_any(&self.pipelines))
+            .load(&mut conn)
+            .await?
+            .into_iter()
+            .collect();
+
+        for pipeline in &self.pipelines {
+            if !latest.contains_key(*pipeline) {
+                return Ok(None);
+            }
+        }
+
+        Ok(latest.into_values().min().map(|l| l as u64))
+    }
+
+    /// Waits until the indexer has caught up to the given checkpoint, or the timeout is reached.
+    pub async fn wait_for_checkpoint(
+        &self,
+        checkpoint: u64,
+        timeout: Duration,
+    ) -> Result<(), Elapsed> {
+        tokio::time::timeout(timeout, async move {
+            loop {
+                if matches!(self.latest_checkpoint().await, Ok(Some(l)) if l >= checkpoint) {
+                    break;
+                } else {
+                    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                }
+            }
+        })
+        .await
+    }
+
+    /// Triggers cancellation of all downstream services, waits for them to stop, and cleans up the
+    /// temporary database.
+    pub async fn stopped(self) {
+        self.cancel.cancel();
+        let _ = self.indexer.await;
+        let _ = self.jsonrpc.await;
```

Comment:
are `_`s here possibly errors, will want to handle them if so

Reply:
`indexer` and `jsonrpc` are `tokio::JoinHandle`s -- they will produce an error if those tasks were aborted or panicked. There's nothing we need to do in those cases if we are trying to wind things down (the panic will have been outputted to the stderr already, see [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2238660cbff9e9918771f18fdb86e671)